src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -98,7 +98,8 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         let intrinsic_name = &self.tcx.item_name(instance.def_id()).as_str()[..];
         match intrinsic_name {
             "caller_location" => {
-                let caller = self.tcx.sess.source_map().lookup_char_pos(span.lo());
+                let topmost = span.ctxt().outer_expn().expansion_cause().unwrap_or(span);
+                let caller = self.tcx.sess.source_map().lookup_char_pos(topmost.lo());
```

Comment:
Would be nice if we could pass `Span` into the query, but @anp ran into trouble with that, I think the result was ICEs from the incremental system.

Reply:
You could just do:
```rust
fn caller(&self, span: Span) -> Whatever {
    let topmost = span.ctxt().outer_expn().expansion_cause().unwrap_or(span);
    self.sess.source_map().lookup_char_pos(topmost.lo())
}
```

Reply:
Where would you put this? Also, "caller" is misleading, in fact the name of `caller_location` is suboptimal (likely due to the RFC assuming an implementation strategy less versatile than the one being followed by @anp).

Reply:
@eddyb I think you are taking the snippet too literally; rename it to anything you see fit. It would be placed on `TyCtxt` -- it could also be placed on `SourceMap`.

Reply:
But, like, there is little generality in it. It's literally "caller_location helper" if it does those two things, and that doesn't feel right to put anywhere outside the query for it (which, again, would ideally take `Span`).

Reply:
I suppose, but it does make sure the logics don't grow apart. I guess we'll have to ensure that through reviews instead.

Reply:
FWIW "you could just do" is my least favorite way to phrase review feedback.

@eddyb is correct that I had ICEs when using a Span as a parameter to a query. I agree with them that two callsites does not a need for deduplication make (yet).